zebra/zebra_nhg.c

```diff
@@ -26,14 +26,1088 @@
+					  nhg_ctx_get_grp(ctx), count)) {
+			depends_decrement_free(&nhg_depends);
+			nexthop_group_delete(&nhg);
+			return -ENOENT;
+		}
+
+		if (!zebra_nhg_find(&nhe, id, nhg, &nhg_depends, vrf_id, type,
+				    afi))
+			depends_decrement_free(&nhg_depends);
+
+		/* These got copied over in zebra_nhg_alloc() */
+		nexthop_group_delete(&nhg);
+	} else
+		nhe = zebra_nhg_find_nexthop(id, nhg_ctx_get_nh(ctx), afi,
+					     type);
+
+	if (nhe) {
+		if (id != nhe->id) {
+			struct nhg_hash_entry *kernel_nhe = NULL;
+
+			/* Duplicate but with different ID from
+			 * the kernel
+			 */
+
+			/* The kernel allows duplicate nexthops
+			 * as long as they have different IDs.
+			 * We are ignoring those to prevent
+			 * syncing problems with the kernel
+			 * changes.
+			 *
+			 * We maintain them *ONLY* in the ID hash table to
+			 * track them.
+			 */
+
+			kernel_nhe = zebra_nhg_copy(nhe, id);
+			zebra_nhg_insert_id(kernel_nhe);
+			zebra_nhg_set_dup(kernel_nhe);
+		} else if (zebra_nhg_contains_dup(nhe)) {
+			/* The group we got contains a duplciate depend,
+			 * so lets mark this group as a dup as well and release
+			 * it from the non-ID hash.
+			 */
+			hash_release(zrouter.nhgs, nhe);
+			zebra_nhg_set_dup(nhe);
+		} else {
+			/* It actually created a new nhe */
+			SET_FLAG(nhe->flags, NEXTHOP_GROUP_VALID);
+			SET_FLAG(nhe->flags, NEXTHOP_GROUP_INSTALLED);
+		}
+	} else {
+		flog_err(
+			EC_ZEBRA_TABLE_LOOKUP_FAILED,
+			"Zebra failed to find or create a nexthop hash entry for ID (%u)",
+			id);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int nhg_ctx_process_del(struct nhg_ctx *ctx)
+{
+	struct nhg_hash_entry *nhe = NULL;
+	uint32_t id = nhg_ctx_get_id(ctx);
+
+	nhe = zebra_nhg_lookup_id(id);
+
+	if (!nhe) {
+		flog_warn(
+			EC_ZEBRA_BAD_NHG_MESSAGE,
+			"Kernel delete message received for nexthop group ID (%u) that we do not have in our ID table",
+			id);
+		return -1;
+	}
+
+	zebra_nhg_handle_kernel_state_change(nhe, true);
+
+	return 0;
+}
+
+static void nhg_ctx_process_finish(struct nhg_ctx *ctx)
+{
+	struct nexthop *nh;
+
+	/*
+	 * Just freeing for now, maybe do something more in the future
+	 * based on flag.
+	 */
+
+	if (nhg_ctx_get_count(ctx))
+		goto done;
+
+	nh = nhg_ctx_get_nh(ctx);
+
+	nexthop_del_labels(nh);
+
+done:
+	if (ctx)
+		nhg_ctx_free(ctx);
+}
+
+static int queue_add(struct nhg_ctx *ctx)
+{
+	/* If its queued or already processed do nothing */
+	if (nhg_ctx_get_status(ctx) == NHG_CTX_QUEUED)
+		return 0;
+
+	if (rib_queue_nhg_add(ctx)) {
+		nhg_ctx_set_status(ctx, NHG_CTX_FAILURE);
+		return -1;
+	}
+
+	nhg_ctx_set_status(ctx, NHG_CTX_QUEUED);
+
+	return 0;
+}
+
+int nhg_ctx_process(struct nhg_ctx *ctx)
+{
+	int ret = 0;
+
+	switch (nhg_ctx_get_op(ctx)) {
+	case NHG_CTX_OP_NEW:
+		ret = nhg_ctx_process_new(ctx);
+		if (nhg_ctx_get_count(ctx) && ret == -ENOENT
+		    && nhg_ctx_get_status(ctx) != NHG_CTX_REQUEUED) {
+			/* Depends probably came before group, re-queue.
```

Comment:
Don't follow what's happening here - what does "depends ... came before group" mean? A block of comments about what the flow is here between the kernel/OS and zebra would be really helpful.

Reply:
This is to account for the possibility of processing a group before any of its depends have been handled. We re-queue once and only once.

Agreed, it is worded terribly, I will add sufficient commenting.